### PR TITLE
fix issue #24

### PR DESCRIPTION
**Pull request by gmarcais**
* panic with valid error message when query missing mates or not name sorted

### CODE REVIEW

File: src/query_bam_records.rs [hunk 1]
```diff
@@ -159,7 +160,7 @@ impl BAMQueryRecordReader {
                     second: second_vec,
                 });
                 i += 1;
-            } else { // paired-end
+            } else if self.record_list[i].is_paired() == true && self.record_list.len() % 2 == 0 { // paired-end
```
Comment (rob-p) [inline src/query_bam_records.rs:163]: Is it necessary to check `self.record_list[i].is_paired() == true` here? Doesn't `is_paried()` itself return a boolean?

**Comment by rob-p** [comment]
We should also see how the solution here interacts with / duplicates / compliments what is going on in #26.

**Comment by gmarcais** [comment]
PR #26 uses the incorrectly mated reads (2 cases: missing mates and mis-ordered mates) as single reads. This PR on the other hand fails with a loud message with an explanation.

The question boils down to what to do when we encounter a poorly formatted BAM: silently handle the data as best as it can, or fail early to avoid unexpected behavior. This PR implements the latest.

Maybe each type of formatting deserves a different treatment?

**Comment by jelber2** [comment]
I had to do the following to get  7aa09c2 to compile (rust complained that there were 2 versions of bio_types [0.12.1 and 0.13.0]. 

```sh
rustup show

Default host: x86_64-unknown-linux-gnu
rustup home:  /nfs/scistore16/itgrp/jelbers/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.62.1 (e092d0b6b 2022-07-16)
```

```sh
git clone https://github.com/OceanGenomics/mudskipper
cd mudskipper/
git fetch origin pull/29/head:29
git checkout 29
git checkout 7aa09c2
perl -pi -e 's/bio-types = "0.12.1"/bio-types = "0.13.0"/' Cargo.toml
cargo build --release
```

**Comment by jelber2** [comment]
I must say that the skip option is handy (if STAR was run with `--outSAMtype BAM Unsorted`) as downstream output might not work with Salmon (mixture of paired and unpaired reads).

**Comment by rob-p** [comment]
Ok; things look good here. I'm going to go ahead with the merge.  Thank you @yfei-w!

**Pull request merged**